pkg/sendbuffer/sendbuffer.go

```diff
@@ -0,0 +1,264 @@
+package sendbuffer
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/kolide/kit/ulid"
+	"github.com/kolide/launcher/pkg/agent/storage/inmemory"
+	"github.com/kolide/launcher/pkg/agent/types"
+)
+
+type sender interface {
+	Send(r io.Reader) error
+}
+
+var (
+	defaultMaxSize     = 128 * 1024
+	defaultMaxSendSize = 8 * 1024
+)
+
+type SendBuffer struct {
+	kvStore                    types.KVStore
+	dataKeys                   [][]byte
+	size, maxSize, maxSendSize int
+	writeMutex, sendMutex      sync.Mutex
+	logger                     log.Logger
+	sender                     sender
+	sendInterval               time.Duration
+	stopSendingChan            chan struct{}
+	isSending                  bool
+}
+
+type option func(*SendBuffer)
+
+func WithMaxSize(maxSize int) option {
+	return func(sb *SendBuffer) {
+		sb.maxSize = maxSize
+	}
+}
+
+func WithMaxSendSize(sendSize int) option {
+	return func(sb *SendBuffer) {
+		sb.maxSendSize = sendSize
+	}
+}
+
+func WithLogger(logger log.Logger) option {
+	return func(sb *SendBuffer) {
+		sb.logger = logger
+	}
+}
+
+func WithKVStore(kvStore types.KVStore) option {
+	return func(sb *SendBuffer) {
+		sb.kvStore = kvStore
+	}
+}
+
+// WithSendInterval sets the interval at which the buffer will send data.
+func WithSendInterval(sendInterval time.Duration) option {
+	return func(sb *SendBuffer) {
+		sb.sendInterval = sendInterval
+	}
+}
+
+func New(sender sender, opts ...option) *SendBuffer {
+	sb := &SendBuffer{
+		maxSize:         defaultMaxSize,
+		maxSendSize:     defaultMaxSendSize,
+		dataKeys:        make([][]byte, 0),
+		sender:          sender,
+		sendInterval:    5 * time.Second,
+		logger:          log.NewNopLogger(),
+		kvStore:         inmemory.NewStore(log.NewNopLogger()),
+		stopSendingChan: make(chan struct{}),
+		isSending:       false,
+	}
+
+	for _, opt := range opts {
+		opt(sb)
+	}
+
+	sb.logger = log.With(sb.logger, "component", "sendbuffer")
+
+	// get all data keys and calculate starting size
+	sb.kvStore.ForEach(func(k, v []byte) error {
+		sb.size += len(k) + len(v)
+		sb.dataKeys = append(sb.dataKeys, k)
+		return nil
+	})
+
+	// sort the keys
+	sort.Slice(sb.dataKeys, func(i, j int) bool {
+		return bytes.Compare(sb.dataKeys[i], sb.dataKeys[j]) < 0
+	})
+
+	return sb
+}
+
+func (sb *SendBuffer) Write(data []byte) (int, error) {
+	sb.writeMutex.Lock()
+	defer sb.writeMutex.Unlock()
+
+	if len(data) == 0 {
+		return 0, nil
+	}
+
+	// if the single data piece is larger than the max send size, drop it and log
+	if len(data) > sb.maxSendSize {
+		sb.logger.Log(
+			"msg", "dropped data because element greater than max send size",
+			"size_of_data", len(data),
+			"max_send_size", sb.maxSendSize,
+			"head", string(data)[0:minInt(len(data), 100)],
+		)
+		return len(data), nil
+	}
+
+	// if we are full, something has backed up
+	// purge everything
+	if len(data)+sb.size > sb.maxSize {
```

Comment:
so this is actually is when we've reached the maximum storage size, I've renamed the var to reflect this. So if we hit this limit it's because were not even attempting to send for some reason.